src/Catalog/Icons/CatalogLeafDataProcessor.cs

```diff
@@ -0,0 +1,274 @@
+            if (hasExternalIconUrl && !hasEmbeddedIcon && Uri.TryCreate(iconUrlString, UriKind.Absolute, out var iconUrl))
+            {
+                using (_logger.BeginScope("Processing icon url {IconUrl}", iconUrl))
+                {
+                    await ProcessExternalIconUrlAsync(destinationStorage, item, iconUrl, cancellationToken);
+                }
+            }
+            else if (hasEmbeddedIcon)
+            {
+                await ProcessEmbeddedIconAsync(destinationStorage, item, iconFile, cancellationToken);
+            }
+        }
+
+        private async Task ProcessExternalIconUrlAsync(IStorage destinationStorage, CatalogCommitItem item, Uri iconUrl, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Found external icon url {IconUrl} for {PackageId} {PackageVersion}",
+                iconUrl,
+                item.PackageIdentity.Id,
+                item.PackageIdentity.Version);
+            if (!IsValidIconUrl(iconUrl))
+            {
+                _logger.LogInformation("Invalid icon URL {IconUrl}", iconUrl);
+                return;
+            }
+            var cachedResult = _iconCopyResultCache.Get(iconUrl);
+            if (cachedResult != null)
+            {
+                if (cachedResult.IsCopySucceeded)
+                {
+                    _logger.LogInformation("Seen {IconUrl} before, will copy from {CachedLocation}",
+                        iconUrl,
+                        cachedResult.StorageUrl);
+                    var tryRegularCopy = false;
+                    var storageUrl = cachedResult.StorageUrl;
+                    var targetStoragePath = GetTargetStorageIconPath(item);
+                    var destinationUrl = destinationStorage.ResolveUri(targetStoragePath);
+                    if (storageUrl == destinationUrl)
+                    {
+                        // We came across the package that initially caused the icon to be added to the cache.
+                        // Skipping it.
+                        return;
+                    }
+                    try
+                    {
+                        await Retry.IncrementalAsync(
+                            async () => await destinationStorage.CopyAsync(storageUrl, destinationStorage, destinationUrl, null, cancellationToken),
+                            e => { _logger.LogWarning(0, e, "Exception while copying from cache {StorageUrl}", storageUrl); return true; },
+                            MaxBlobStorageCopyAttempts,
+                            initialWaitInterval: TimeSpan.FromSeconds(5),
+                            waitIncrement: TimeSpan.FromSeconds(1));
+                    }
+                    catch (Exception e)
+                    {
+                        _logger.LogWarning(0, e, "Copy from cache failed after {NumRetries} attempts. Falling back to copy from external URL. {StorageUrl}",
+                            MaxBlobStorageCopyAttempts,
```

Comment:
This case seems unlikely, does this happen? Just curious

Reply:
There is no proper cache cleanup logic when package is deleted, so we get here when there was a package delete that removes the icon. I haven't seen any other failures that got us here.